### ISSUE TITLE
ecsv round trip failure

### ISSUE THREAD

**Issue by keflavich** [astropy/astropy#4415]
I've encountered a round-trip failure using ecsv files:

```
from astroquery.vizier import Vizier
from astropy.table import Table

Vizier.ROW_LIMIT = 1e5
moxc = Vizier.query_constraints(catalog='J/ApJS/213/1/table3', GLON='>49.0 & < 51.0',)
w51moxc = moxc[0][(moxc[0]['Region'] == b'W51A')]

w51moxc.write('w51_moxc.ecsv', format='ascii.ecsv')
Table.read('w51_moxc.ecsv', format='ascii.ecsv')
```

resulting error:
```
## -- End pasted text --
Traceback (most recent call last):
  File "<ipython-input-1-c25aaff5a254>", line 9, in <module>
    Table.read('w51_moxc.ecsv', format='ascii.ecsv')
  File "/Users/adam/repos/astropy/astropy/table/table.py", line 2261, in read
    return io_registry.read(cls, *args, **kwargs)
  File "/Users/adam/repos/astropy/astropy/io/registry.py", line 332, in read
    data = reader(*args, **kwargs)
  File "/Users/adam/repos/astropy/astropy/io/ascii/connect.py", line 37, in io_read
    return read(filename, format=format, **kwargs)
  File "/Users/adam/repos/astropy/astropy/io/ascii/ui.py", line 339, in read
    dat = reader.read(table)
  File "/Users/adam/repos/astropy/astropy/io/ascii/core.py", line 1069, in read
    raise InconsistentTableError(errmsg)
InconsistentTableError: Number of header columns (56) inconsistent with data columns (55) at data line 0
Header values: ['_RAJ2000', '_DEJ2000', 'Region', 'p', 'CXOU', 'Label', 'RAJ2000', 'DEJ2000', 'ePos', 'PosType', 'no-m', 'no-t', 'no-s', 'no-h', 'KS_S', 'KS_M', 'ExpNom', 'ExpFrac', 'Ct', 'Nobs', 'Nph', 'Fdis', 'b_theta', 'theta', 'B_theta', 'PSFf', 'Area', 'aglow', 'Oct-t', 'Oct-s', 'Oct-h', 'bgScl', 'bgct-t', 'bgct-s', 'bgct-h', 'b_Nct-t', 'Nct-t', 'B_Nct-t', 'b_Nct-s', 'Nct-s', 'B_Nct-s', 'b_Nct-h', 'Nct-h', 'B_Nct-h', 'Area-t', 'Area-s', 'Area-h', 'E-t', 'E-s', 'E-h', 'XFlx-t', 'XFlx-s', 'XFlx-h', 'EFlx-t', 'EFlx-s', 'EFlx-h']
Data values: ['290.900776', '14.469873', 'W51A', '192336.18+142811.5', 'p1_253', '290.900776', '14.469873', '0.43', 'AE mean data', '0.0006', '0.0074', '0.0006', '--', '--', '--', '70214.7', '0.96', '0.00015', '2', '2', '0.00', '3.23', '3.23', '3.23', '0.89', '20.16', '--', '3', '3', '0', '305.1', '125', '48', '77', '1.1', '2.6', '4.7', '1.4', '2.8', '5.0', '--', '-0.3', '1.1', '225.06', '323.32', '200.43', '1.23', '1.23', '--', '1.6e-07', '1.2e-07', '-1.8e-08', '3e-16', '3e-16', '--']
```

**Comment by taldcroft** [comment]
Fundamentally this is due to the column `p` that has all blank values, in conjunction with a behavior of the `csv` module writer that surprises me:
```
In [35]: writer = csv.writer(sys.stdout, quotechar='"', delimiter=' ')
In [36]: n_out = writer.writerow([' '] * 3)
" " " " " "
Out[36]:
In [37]: n_out = writer.writerow([''] * 3)
  
Out[37]: 
```
In other words the csv.writerow() function does not quote a completely empty string.  This is OK if you are literally splitting on a single space character later when reading, but in our case (and in most CSV parsers) you allow for any contiguous whitespace as the delimiter.

A workaround is to use `,` (comma) as the delimiter.  It turns out for a masked table such as you have, in order to read it back you need to use something like:
```
dat = Table.read('w51_moxc.ecsv', format='ascii.ecsv', fill_values=['--', '0'])
```

A final issue is that the table will come back with string instead of bytestring columns in Py3.  The current ECSV specification does not have a type for bytestring [1].  At some level this is intentional, though of course it could be open for discussion.  

In your particular use case the "natural" behavior would be that all of the string columns returned by `query_constraints` would be strings and not bytestrings.  This is reflected in the general "unicode sandwich" idea of making everything on the "inside" that the user handles be strings while the outside (the file) may be bytestrings as needed.

[1] - https://github.com/astropy/astropy-APEs/blob/master/APE6.rst#header-details

**Comment by taldcroft** [comment]
Fixed by #4417.

**Comment by embray** [comment]
Why does the table returned from astroquery contain bytes strings?  Does it not up-convert string columns to unicode strings?
